hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/server/namenode/ha/ConfiguredFailoverProxyProvider.java

```diff
@@ -93,4 +109,65 @@ public synchronized void close() throws IOException {
   public boolean useLogicalURI() {
     return true;
   }
+
+  /**
+   * Write active NameNode index to cache file.
+   */
+  public void writeActiveCache(int index) {
+    if (!cacheActiveEnabled) {
+      return;
+    }
+
+    synchronized (cacheActiveFile) {
+      boolean exist = cacheActiveFile.exists();
+      try (RandomAccessFile raf = new RandomAccessFile(cacheActiveFile, "rw");
+          FileChannel fc = raf.getChannel();
+          FileLock lock = fc.tryLock(0, Long.MAX_VALUE, false)) {
+        if (lock != null) {
+          raf.setLength(0);
+          raf.writeBytes(String.valueOf(index));
+          if (!exist) {
+            boolean ret = cacheActiveFile.setWritable(true, false)
+                && cacheActiveFile.setReadable(true, false);
+            if (!ret) {
+              throw new IOException("Cannot set file rw mode.");
+            }
+          }
+          LOG.debug("Succeed in writing active index " + index
+              + " to cache file " + cacheActiveFile);
+        }
+      } catch (Throwable e) {
+        LOG.warn("Filed to write active index " + index + " to cache file "
+            + cacheActiveFile, e);
```

Comment:
how about adding even more hints to let user know how to troubleshoot it? For example, tell user to configure dfs.client.failover.cache-active.dir

Reply:
OK, thanks, I have added it like this:

```
2019-11-25 14:54:32,791 WARN ha.AbstractNNFailoverProxyProvider: Failed to write active index 1 to cache file /tmp/ns1, It may be due to permission issues, the dfs.client.failover.cache-active.dir must be properly configured!
java.io.FileNotFoundException: /tmp/ns1 (Permission denied)
	at java.io.RandomAccessFile.open0(Native Method)
	at java.io.RandomAccessFile.open(RandomAccessFile.java:316)
	at java.io.RandomAccessFile.<init>(RandomAccessFile.java:243)
```